OpenAMP support.

iabdalkader opened the pull request:
This PR adds support for OpenAMP in MicroPython, with the goal of standardizing asymmetric multi-core support across different hardware running MicroPython. There are 3 major components to OpenAMP, `libmetal`, `remoteproc` and `RPMsg`. `libmetal` provides abstraction of the low-level underlying hardware, `remoteproc` is used for processor Life Cycle Management (LCM) (i.e loading firmware, starting, stopping a core etc..), and `RPMsg` is a bus infrastructure that enables Inter Processor Communications (IPC) between different cores. These components are implemented in this PR as follows:

* OpenAMP's `libmetal` submodule is added in this PR (see note about this submodule), and a generic MicroPython system port is implemented, one that uses some sane defaults and common `mp_hal` functions, and also allows for MicroPython's ports to override defaults and config `libmetal`.

* OpenAMP's `remoteproc` component is implemented in `extmod/modremoteproc.c` module which provides an API to load firmware and control remote processors. Port-specific ops must be implemented for this module (similar to `mpbthciport.c` or `mpnetworkport.c`). One example is already done in this PR for `stm32`.

* OpenAMP's `RPMsg` component and the rest of the logic is implemented in `extmod/modopenamp.c`. This module provides an`endpoint` type (a logical connection on top of `RPMsg` channel) that can be used to communicate with the remote core.

Example: The following script loads a firmware image that implements a virtual uart over RPMsg channel, built for STM32H7's M4 core, starts it, and communicates with the M4 once the core is ready to communicate. The remote core becomes ready to communicate after the master/host core initializes a shared memory area, and it announces its readiness by creating a new RPMsg channel. At this point, the `ns_callback` is called, and the host processor creates an endpoint to communicate over that RPMsg channel. Note the firmware is parsed and loaded from an `elf` file stored on the filesystem, I can share the source somewhere later as an example or for testing.

```Python
import openamp
import time

ept = None


def ept_recv_callback(src_addr, data):
    print(f'Received message on endpoint "{data}"')


def ns_bind_callback(src_addr, name):
    global ept
    print(f'New service announcement src address: {src_addr} name: "{name}"')
    # Create a new RPMsg endpoint to communicate with the M4.
    ept = openamp.RPMsg(
        name, src_addr=openamp.RPMSG_ADDR_ANY, dst_addr=src_addr, callback=ept_recv_callback
    )


# Initialize OpenAMP. Name service callback is called when the remote
# processor creates an RPMsg channel, and is ready to communicate.
openamp.init(ns_callback=ns_bind_callback)

# Create a remoteproc object, load its firmware and start it.
# Note you can also start from an entry point (ex 0x081E0000)
#rproc = openamp.RProc(0x081E0000)
rproc = openamp.RProc("vuart.elf")
rproc.start()

while True:
    if ept is not None:
        ept.send("Hello World!", timeout=1000)
    time.sleep_ms(1000)
```

Output:
```bash
registered generic bus
remoteproc_load: check remoteproc status
remoteproc_load: open executable image
remoteproc_load: check loader
remoteproc_load: loading headers
Loading ELF headering
Loading ELF program header.
Loading ELF section header.
remoteproc_load, load header 0x0, 0x1000, next 0x18098, 0x258
Loading ELF section header.
Loading ELF section header complete.
Loading ELF shstrtab.
remoteproc_load, load header 0x18098, 0x258, next 0x1800d, 0x8a
Loading ELF shstrtab.
remoteproc_load, load header 0x1800d, 0x8a, next 0x1800d, 0x0
remoteproc_load: load executable data
segment: 1, total segs 1
load data: da 0x38000000, offset 0x10000, len = 0x7f90, memsize = 0xc138, state 0x20801
load data: da 0xffffffff, offset 0x0, len = 0x0, memsize = 0x0, state 0x40801
remoteproc_load: successfully load firmware
New service announcement src address: 1024 name: "m4-vuart-channel"
Received message on endpoint "Hello from M4!"
```

The firmware running on the M4 can be found here: https://github.com/iabdalkader/openamp_vuart

Todo:

- [x] Handle file system access in a generic way using mp_vfs
- [x] Support endpoint send/send non-blocking and off-channel.
- [x] Add optional trace buffer resource.
- [x] Support loading/starting firmware from Flash address.
- [x] Support loading/starting firmware from SD-RAM on H7.
- [x] Provide an example for the H7 M4 side (optional).


* Note about `libmetal` it's basically impossible to add its upstream as a submodule, since it only supports cmake, and it hard-codes `configure_file` variables into headers, and it's basically expected to be built into a static library and installed somewhere. So for now I've used my own fork, with minor changes to make it more "MicroPython-friendly". This library hardly changes, I think it would make sense to maintain our own fork, I can transfer ownership of the `libmetal` repo to MicroPython later if you want, I have no need to maintain it.

iabdalkader committed:
lib/open-amp: Add OpenAMP submodule.

OpenAMP framework provides a standard inter processor communications
infrastructure for RTOS and bare metal environments. There are 3 major
components in OpenAMP: libmetal, remoteproc and RPMsg. libmetal provides
abstraction of the low-level underlying hardware, remoteproc is used for
processor Life Cycle Management (LCM) like loading firmware, starting,
stopping a core etc., and RPMsg is a bus infrastructure that enables Inter
Processor Communications (IPC) between different cores.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32: Implement port backend for libmetal.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32: Implement port backend for OpenAMP's remoteproc.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_GIGA: Enable OpenAMP.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
py/stream: Add mp_stream_seek() helper function.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented on ports/stm32/mprprocport.c:
```diff
@@ -0,0 +1,253 @@
+    mp_obj_t file;
+} filestore_t;
+
+static int store_open(void *store, const char *path, const void **image_data) {
```
It seems like this function, and the one below (and others in this file maybe) are generic enough that they could be put in `extmod/modrproc.c`?

iabdalkader replied:
They are now indeed, thanks to switching to the VFS functions. I can move them to `modrproc.c` as a generic file store enabled/disabled via config (ex `MICROPY_PY_OPENAMP_FILE_STORE`) so ports can still override them (if needed) by providing their own `image_store_ops`.

iabdalkader replied:
Done. I think it would be nice if the file store is moved to a separate file, ex `modrproc_store.c`, where maybe other stores can be added.

iabdalkader commented:
@dpgeorge 

It would be really nice if we can update the CMSIS headers to support zero and copy tables, this way I can remove the function that initializes the shared resource table (`openamp_rsc_table_init`) .

https://github.com/ARM-software/CMSIS_5/blob/a75f01746df18bb5b929dfb8dc6c9407fac3a0f3/CMSIS/Core/Include/cmsis_gcc.h#L137

dpgeorge commented:
> It would be really nice if we can update the CMSIS headers to support zero and copy tables

We can.  But not until after the upcoming release, which will be in the next couple of weeks.

iabdalkader committed:
stm32/mpu: Add MPU config for shared, uncached memory region.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
> > It would be really nice if we can update the CMSIS headers to support zero and copy tables
> 
> We can. But not until after the upcoming release, which will be in the next couple of weeks.

I'm still waiting on the feedback, but in the meantime, I was wondering if now would be a good time to update CMSIS to support for copy/zero tables ? It would help with simplifying some of the code here, but might also be required for other future ports.

jimmo reviewed (commented):
Thanks @iabdalkader for all the hard work on this! We're keen to go forward with this general direction with OpenAMP as the supported mechanism for multi-core (and specifically AMP) microcontrollers.

We had a look at other implementations of OpenAMP, and in particular Zephyr. At the end of the day, the API we want to provide should allow simple connection of channels/endpoints, and a way to send/receive messages on those channels.

To that end, Zephyr has done a very good job of providing a high-level API that does exactly this: https://docs.zephyrproject.org/apidoc/latest/group__rpmsg__service__api.html (although of course you can still use OpenAMP directly from Zephyr too).

Implementation here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/ipc/rpmsg_service/rpmsg_service.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/ipc/rpmsg_service.h

This would mean moving the implementation of e.g. `ns_bind_callback` into C, with the expectation that a standard one-size-fit-all way of matching up endpoints would for most use cases.

We note that Zephyr chooses the master/slave role statically. Eventually it would be nice to make this dynamic (and that doesn't seem too difficult) but for now it seems reasonable to make it static at compile time.

So the example in https://github.com/micropython/micropython/pull/12961#issue-1991008119 would become:

```py
import openamp
import time

def ept_recv_callback(src_addr, data):
    print(f'Received message on endpoint "{data}"')

ept = openamp.Endpoint("m4-vuart-channel", ept_recv_callback)

rproc = openamp.RProc("vuart.elf")
rproc.start()

while True:
    if ept.isconnected():
        ept.send("Hello World!", timeout=1000)
    time.sleep_ms(1000)
```

The overall point is that all you need to know is the name of the endpoint and they connect automatically.

But really, what we'd love to see is that the `Endpoint` implement the stream API (exactly like a UDP socket). This would make this directly usable from asyncio. The stream would become writable once connected.

This would lose the ability to have a callback message received, however:
* Programs should already be using select/poll to consume events from multiple source.
* Or better, using asyncio.
* We might decide to make a formal "add callback to stream" API that works on all stream types in the future.

Happy to discuss further before you make any changes, let us know.

iabdalkader commented on .gitmodules:
```diff
@@ -59,3 +59,9 @@
 [submodule "lib/protobuf-c"]
 	path = lib/protobuf-c
 	url = https://github.com/protobuf-c/protobuf-c.git
+[submodule "lib/libmetal"]
```
I can fork the repo and redo the changes on top of the stable release tag, then replace the submodule that's easy to do, but I don't think we can avoid modifying the library in its current state. Maybe with small upstream changes it would be possible, for example allow us to provide an external config header. Also we can "pretend" to be the "generic' system, then what they already have here works for us:

```c
#if defined(HAVE_PROCESSOR_CPU_H)
# include <metal/processor/@PROJECT_PROCESSOR@/cpu.h>
#else
# include <metal/processor/generic/cpu.h>
#endif
```

But then some other headers include this unconditionally:

```
#include <metal/system/@PROJECT_SYSTEM@/io.h>
```
 
That would need to be fixed too. I think can take another look at this and try to make it work with minimal changes, then possibly submit those upstream.

iabdalkader commented:
Hi @jimmo 

Thanks for the thorough review! I'll get around to addressing all of the review comments next week, when I get back from vacation.

> This would mean moving the implementation of e.g. ns_bind_callback into C, with the expectation that a standard one-size-fit-all way of matching up endpoints would for most use cases.

I think we can have both, I need to double check to be sure, but we can either create the channel in the announcement callback, or have it created beforehand. In the latter case, it will remain unconnected until the remote opens the channel, then open-amp will match it. In the C callback I can check if the Python callback is `None`, then assume the channel is already created, note I may not even have to do that as it may not call the callback in that case. Anyway, if this is the case, I assume there's no problem in keeping the Python `ns_bind_callback` option, right ?

> We note that Zephyr chooses the master/slave role statically. Eventually it would be nice to make this dynamic (and that doesn't seem too difficult) but for now it seems reasonable to make it static at compile time.

I've only added support for Master/Host mode, there's no support for a remote role, not even statically, because to target the remote cores, you'd first have to support building MicroPython for the M4 core of the STM32H7 or the RT1170 for example, so I left this as an exercise for the reader :)

> To that end, Zephyr has done a very good job of providing a high-level API

I haven't looked at the Zephyr implementation, but the general idea here might be the opposite. I wanted to provide a sort of "low-level" Open-AMP Python API, on top of which you can build more complex Python libraries, not something that's high-level. For example I use the current API in a MSGPack RPC module that supports sync/async remote function calls, it's unpublished yet, but here's an example usage, you can see how it wraps the API details:

```Python
    rpc = msgpackrpc.MsgPackRPC()

    # Register objects or functions to be called by the remote processor.
    rpc.bind(function_to_called_by_remote)

    # Start the remote processor and wait for it to be ready to communicate.
    rpc.start(firmware=0x08180000, timeout=1000, num_channels=2)

    # Sync call
    print(rpc.call("add", 1, 1))

    # Async call
    future = rpc.call_async("add", 1, 2)
    print(future.join())
```

That said, I understand that there might be simple use cases in which a send/recv data in a loop is all that's needed, but I think we can still have that without sacrificing the flexibility of the current API.

> But really, what we'd love to see is that the Endpoint implement the stream API (exactly like a UDP socket). This would make this directly usable from asyncio. The stream would become writable once connected.

Note that there's no `rpmsg_recv` function you basically get the buffer and its size in the `recv_callback` and then you're expected to consume it immediately. That said, I think we can still have both here too. If the callback is None, buffer the message in a FIFO and return it in a recv call or stream read (ex. "streaming mode"), but is this something that has to be implemented right away or can it be a later improvement ?

> Or better, using asyncio.

It's important to say that I intentionally didn't want this to have a hard-dependency on `asyncio` because then everything we write that uses it has to be `asyncio` code, but having an `asyncio` *option* is fine though, as long as we don't have to use asyncio to use open-amp.

iabdalkader commented on extmod/modrproc.c:
```diff
@@ -0,0 +1,218 @@
+
+#else
+// A generic file-based image store.
+static int store_open(void *store, const char *path, const void **image_data) {
```
+1 for sure.

projectgus commented on ports/stm32/Makefile:
```diff
@@ -417,6 +417,13 @@ endif
 endif # MICROPY_PY_BLUETOOTH
 
+ifeq ($(MICROPY_PY_OPENAMP),1)
+SRC_C += metal/metal_port.c
```
Suggest this directory is renamed `openamp` (or `openamp/metal`), even though it currently only contains a metal port.

The reason is to avoid developers needing to be familiar with two openamp-associated names at the port level. From our perspective `metal` is part of the implementation of `openamp`, so it's easier to understand it if it's scoped there first. (Currently most of the "high level" names are `openamp` already, I think it's just this metal/ subdirectory.)

This may seem like a nitpick, but Micropython is a big project so it's useful to keep the naming clean whenever possible.

iabdalkader replied:
No it's good suggestion, if you want to override something else or maybe provide a new image store, this will be the right place to add it.

jimmo replied:
To add some more context on this... my understanding was that we only created subdirectories in ports when need to match an existing project's include path. i.e. if the third-party project does `#include "foo/config.h"` then we need do be able to pass something to `-I` that contains `foo`. (e.g. lwip, which expects to be able to do `#include "arch/cc.h", which is why we "hide" it inside the "lwip_inc" dir).

In this case, we actually control the include, so it can just be "mpmetalport.h" as in the commit below (or "mpopenampport_libmetal.h" might be better).

iabdalkader replied:
This is why I originally added it to `metal/metal_config.h` because port dir is included in the path by default, and the library files included `metal/x.h` so I wanted to match the include paths in the library. I prefer `ports/<port>/mpmetalport.h/c` just to avoid making the file name any longer than necessary.

projectgus replied:
Fair enough, thanks @jimmo for the extra explanation.

iabdalkader commented on ports/stm32/metal/metal_port.h:
```diff
@@ -0,0 +1,71 @@
+#define METAL_SHM_SIZE              ((size_t)(_openamp_shm_region_end - _openamp_shm_region_start - METAL_RSC_SIZE))
+
+#define MTEAL_MPU_REGION_BASE       ((uint32_t)_openamp_shm_region_start)
+#define MTEAL_MPU_REGION_SIZE       (MPU_REGION_SIZE_64KB)
```
I'm still missing it :) where is the typo ?

jimmo replied:
`MTEAL`

jimmo commented:
Thanks for the update @iabdalkader -- I will reply with more comments, but wanted to specifically address the libmetal submodule first.

Here's an attempt at a very basic implementation of their pre-processor that allows us to use the upstream submodule. It also makes it much clearer what the minimal diff is to the "micropython" system (compared to the "generic" system that it's based on) as only three files change.

https://github.com/micropython/micropython/commit/52eef33d3e9dee981e590234ea77fcf5e14a87b9

Further to what @projectgus was saying about the directory/file naming of `metal_port.c` -- additionally the port-specific stuff should match the existing convention and use "mpXYZport.c". So I've changed it to "ports/stm32/mpmetalport.h". Perhaps this should be `mpopenampport_libmetal.{h,c}` to match what Angus was suggesting. (This sort of matches what we already do e.g. `unix/mpbtstackport_h4.c`).

iabdalkader commented:
> additionally the port-specific stuff should match the existing convention and use "mpXYZport.c". 

I was following how `mbedtls` is ported (i.e. `mbedtls/mbedtls_config.h`, `mbedtls/mbedtls_port.c`) but I'm okay with `mpmetalport.h`, I don't think it should be any longer, will rename them.

I will add your changes to libmetal, thanks!

iabdalkader committed:
lib/libmetal: Add libmetal submodule.

libmetal provides an abstraction of the underlying hardware, to support
other OpenAMP components.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
extmod/libmetal: Add MicroPython platform for libmetal.

Add a MicroPython platform for libmetal, based on the generic platform.
The MicroPython platform uses common mp_hal_xxx functions and allows ports
to customize default configurations for libmetal.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>
Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
@jimmo 

> Here's an attempt at a very basic implementation of their pre-processor that allows us to use the upstream submodule. It also makes it much clearer what the minimal diff is to the "micropython" system (compared to the "generic" system that it's based on) as only three files change.
> 
> [52eef33](https://github.com/micropython/micropython/commit/52eef33d3e9dee981e590234ea77fcf5e14a87b9)

You missed a few files, for example https://github.com/OpenAMP/libmetal/compare/main...iabdalkader:libmetal:v2023.10.0-micropython#diff-1def8da686dc921f56f400212059bf90033737f5d7928085c7bd98a5a278c188

I'm going to add them back, did you remove them for a reason ?

jimmo commented:
> You missed a few files, for example [OpenAMP/libmetal@main...iabdalkader:libmetal:v2023.10.0-micropython#diff-1def8da686dc921f56f400212059bf90033737f5d7928085c7bd98a5a278c188](https://github.com/OpenAMP/libmetal/compare/main...iabdalkader:libmetal:v2023.10.0-micropython#diff-1def8da686dc921f56f400212059bf90033737f5d7928085c7bd98a5a278c188)
> 
> I'm going to add them back, did you remove them for a reason ?

@iabdalkader There's no diff from the generic version there.

The steps I took were:
- Start with the version in https://github.com/iabdalkader/libmetal/tree/v2023.10.0-micropython
- Undo the `@PLATFORM...@` change
- Diff each file in plaform/micropython with platform/generic
- Only keep the modified files. (With the exception of log.h which could be done directly in `mpmetalport.h`)

iabdalkader commented:
Oh okay, so it falls back to the generic version.

iabdalkader committed:
extmod/modopenamp: Add new OpenAMP module.

This module implements OpenAMP's basic initialization and shared resources
support, and provides support for OpenAMP's RPMsg component, by providing
an `endpoint` type (a logical connection on top of RPMsg channel) which can
be used to communicate with the remote core.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
extmod/modopenamp_remoteproc: Add new OpenAMP RemoteProc class.

RemoteProc provides an API to load firmware and control remote processors.

Note: port-specific operations must be implemented to support this class.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
@jimmo
> Only keep the modified files. (With the exception of log.h which could be done directly in mpmetalport.h)

It doesn't build with that change, because headers include `log.h` before the `mpmetalport.h` `undef`s it. I reverted the logging and confirmed that it is working again.

projectgus reviewed (commented):
Noticed one more small naming thing, but I didn't see anything else aside from what's already in discussion. This is going to be a very interesting feature!

iabdalkader commented:
I've pushed the branch again to reword the outdated commit messages, and add a config file include for open-amp. If needed the config file will allow ports/boards to override openamp/rproc options, provide a different resource table etc.. It should be defined in `mk` files following mbedtls and others:
```c
MICROPY_PY_OPENAMP_CONFIG_FILE = '"$(BOARD_DIR)/openamp_config_board.h"'
```

I'm very happy with the current state of this feature, but let me know if there's something else to be fixed while we're at it. Note I don't know how/where to add the coverage test for stream seek function.

iabdalkader commented on extmod/extmod.mk:
```diff
@@ -514,3 +515,90 @@ include $(TOP)/extmod/btstack/btstack.mk
+INC += -I$(TOP)/$(OPENAMP_DIR)
+CFLAGS += -DMICROPY_PY_OPENAMP=1
+
+# libmetal is intended to run through a pre-processor (as part of its CMake-based build system).
```
Can we move this block (all the sed lines and SRC_LIBMETAL_C etc..) to `extmod/libmetal/metal.mk` ?

jimmo replied:
Yes that would be good.

dpgeorge replied:
Yes, that sounds fine.  All things related to libmetal configuration can go there.

iabdalkader replied:
Done.

jimmo commented on extmod/libmetal/metal/system/micropython/alloc.h:
```diff
@@ -0,0 +1,40 @@
+#error "Include metal/alloc.h instead of metal/micropython/alloc.h"
+#endif
+
+#ifndef __METAL_MICROPYTHON_ALLOC__H__
```
I noticed you've changed these from the generic versions. My intention was to have as few diffs as possible, and to annotate each line that was changed with a `// Micropython...` comment.

I talked to @dpgeorge about this, and we think there's a simpler solution that means we don't have to copy any boilerplate, and in particular avoids duplicating or having to introduce a new license header.

1. Replace each of `alloc.h`, `log.h`, `sleep.h`, `sys.h` with a file that just contains a single line:

```c
#include <metal/micropython.h>
```

3. Then `extmod/libmetal/metal/micropython.h` contains the functional bits of each of these four files, with none of `#ifdefs`, `extern "C"`, copyright headers, etc.

```c
// MicroPython copyright header...

#ifndef MICROPY_INCLUDED_METAL_MICROPYTHON_H
#define MICROPY_INCLUDED_METAL_MICROPYTHON_H

#include <stdlib.h>
#include <metal/config.h>

// generic/alloc.h
static inline void *__metal_allocate_memory(unsigned int size) {
    // MicroPython: changed from "malloc()"
    return m_tracked_calloc(1, size);
}

static inline void __metal_free_memory(void *ptr) {
    // MicroPython: changed from "free()"
    m_tracked_free(ptr);
}

// generic/log.h
#if METAL_LOG_HANDLER_ENABLE
#include "py/mphal.h"
#undef metal_log
#define metal_log(level, ...) mp_printf(&mp_plat_print, __VA_ARGS__)
#endif

// generic/sleep.h
// Must be provided by the port (in mpmetalport.h / mpmetalport.c).
int __metal_sleep_usec(unsigned int usec);

// sys.h
#ifndef METAL_MAX_DEVICE_REGIONS
#define METAL_MAX_DEVICE_REGIONS 1
#endif

struct metal_state {
    struct metal_common_state common;
};

#ifdef METAL_INTERNAL

// generic/template/sys.h
void sys_irq_enable(unsigned int vector);
void sys_irq_disable(unsigned int vector);

// generic/sys.h
void sys_irq_restore_enable(unsigned int flags);
unsigned int sys_irq_save_disable(void);
#endif

#endif MICROPY_INCLUDED_METAL_MICROPYTHON_H
```

iabdalkader replied:
Done, but we can just add this to `config.h` ?

iabdalkader replied:
After doing this change, it broke the build because `sys.h` includes `metal/micropython.h` before it defines `struct metal_common_state`. I had to move the struct back to `system/micropython/sys.h`

```
In file included from build-ARDUINO_GIGA/openamp/metal/system/micropython/log.h:1,
                 from build-ARDUINO_GIGA/openamp/metal/log.h:114,
                 from build-ARDUINO_GIGA/openamp/metal/sys.h:18,
                 from ../../extmod/modopenamp.c:34:
build-ARDUINO_GIGA/openamp/metal/micropython.h:63:31: error: field 'common' has incomplete type
   63 |     struct metal_common_state common;
```

iabdalkader replied:
I moved everything to `config.h`.

jimmo commented on ports/stm32/mprprocport.c:
```diff
@@ -0,0 +1,154 @@
+#include "metal/utilities.h"
+#include "extmod/modopenamp_rproc.h"
+
+#define debug_printf(...)   // mp_printf(&mp_plat_print, __VA_ARGS__)
```
`DEBUG_printf`

jimmo commented on py/stream.c:
```diff
@@ -82,6 +82,18 @@ mp_uint_t mp_stream_rw(mp_obj_t stream, void *buf_, mp_uint_t size, int *errcode
     return done;
 }
 
+mp_uint_t mp_stream_seek(mp_obj_t stream, mp_off_t offset, int whence, int *errcode) {
```
Please add a coverage test for this.

jimmo replied:
Sorry I missed your other comment 

> Note I don't know how/where to add the coverage test for stream seek function.

`ports/unix/coverage.c` is where we add tests for the C API, there are already some stream examples there.

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+extern mp_obj_type_t rproc_type;
+#endif
+
+static struct metal_device shm_device = {
```
This (and in other places in this file) should be `STATIC` but we've decided it's finally time to get rid of that (see #13763) so actually before this gets merged we'll need to replace all the `STATIC` with `static` in this PR. (Sorry about that! #13763 has a script you can run)

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+
+MP_DEFINE_CONST_OBJ_TYPE(
+    virtio_dev_type,
+    MP_QSTR_virtio_dev,
```
This is a Python class so should be `MP_QSTR_VirtIODev`

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+
+MP_DEFINE_CONST_OBJ_TYPE(
+    rpmsg_type,
+    MP_QSTR_rpmsg,
```
`MP_QSTR_RPMsg` to match the locals dict below.

But I think this class should be named `Endpoint` as per https://github.com/micropython/micropython/pull/12961#pullrequestreview-1869082227

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_rpmsg) },
+    { MP_ROM_QSTR(MP_QSTR___del__), MP_ROM_PTR(&rpmsg_deinit_obj) },
+    { MP_ROM_QSTR(MP_QSTR_send), MP_ROM_PTR(&rpmsg_send_obj) },
+    { MP_ROM_QSTR(MP_QSTR_is_ready), MP_ROM_PTR(&rpmsg_is_ready_obj) },
```
Although I prefer `is_ready` too, the convention elsewhere in MicroPython would be `isready`. (e.g. wlan.isconnected()`.

iabdalkader replied:
I will change it if we must, but I'd argue that it's `isconnected` that's wrong and should be made obsolete, with a non-breaking change by just adding an alias `is_connected`. Not in MicroPython there's `i2c.is_ready` and `is_connected` in bluetooth stuff.
 CC @dpgeorge

dpgeorge replied:
> and `is_connected` in bluetooth stuff

Where is this exactly?

> I'd argue that it's `isconnected` that's wrong and should be made obsolete, with a non-breaking change by just adding an alias `is_connected`

Yes, we could do that.

Are there any other places where the convention "without underscore" is used that would also need changing?  Eg something like `onxxx`.

iabdalkader replied:
> Where is this exactly?

In `micropython-lib` aioble.

> Are there any other places where the convention "without underscore" is used

It's hard to find them, as they match a lot of things, but math functions `isinf` etc.. I don't know if you want to change those or not.

dpgeorge replied:
> In `micropython-lib` aioble.

Ah, I see, thanks.

> It's hard to find them, as they match a lot of things, but math functions `isinf` etc

They can't be changed, that's CPython (and ultimately C library) legacy.  (Although note that CPython has things like `pathlib.Path.is_dir`, so isn't fully consistent itself.)

I don't think in MicroPython's custom modules (eg `machine`, `network`) that we have many of these cases.  For example, in the `micropython` module all the functions there have `_` to separate words, eg `stack_use`, `heap_lock`.  There are things like `readinto` but they also have history from CPython.

If `isconnected()` is the only case that needs to be changed to make things semi-consistent then maybe it's worth doing (in a backwards compatible way), and then this here can be `is_ready`.

iabdalkader replied:
>> If isconnected() is the only case that needs to be changed to make things

`isconnected()` is the only case I could find and it seems to be common in wlan modules. There's already `is_ready` but no `isready` so I think this case here should definitely stay `is_ready`

jimmo commented on extmod/modopenamp_rproc.c:
```diff
@@ -0,0 +1,161 @@
+        if (error != 0) {
+            mp_raise_OSError(error);
+        }
+        #endif
```
Should this have an `#else` to raise (I think) `TypeError`? (i.e. expecting an `int` because we only support the address form, although I guess you could make a case for `ValueError`?)

iabdalkader replied:
For now yes, but eventually the `#else` should then try to load a `.bin` if elf loading is disabled.

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+    return 0;
+}
+
+STATIC mp_obj_t _rpmsg_send(uint n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
```
Why does this need the leading underscore? I'm guessing it's conflicting with a symbol in OpenAMP. Technically leading underscore names are reserved, so we avoid using them.

Perhaps we should prefix all the symbols in this file with `modopenamp_` so that everything is consistent.

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+        }
+    }
+
+    virtio_dev_obj_t *self = mp_obj_malloc_with_finaliser(virtio_dev_obj_t, &virtio_dev_type);
```
It's a bit unconventional to call this `self` which would normally be used for a method on a class. (I was confused reading this because it made me think that I was inside an RPMsg function).

I think it would be clearer if it just used `MP_STATE_PORT(virtio_device)` everywhere in place of `self`.

jimmo commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,394 @@
+STATIC const mp_rom_map_elem_t globals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_openamp) },
+    { MP_ROM_QSTR(MP_QSTR_RPMSG_ADDR_ANY), MP_ROM_INT(RPMSG_ADDR_ANY) },
+    { MP_ROM_QSTR(MP_QSTR_init), MP_ROM_PTR(&openamp_init_obj) },
```
We don't really have a precedent for module-level `init()` elsewhere in MicroPython.

I wonder if we should just make it init-on-demand (i.e. when you first construct an endpoint), and then we'd just need a way to set the ns callback at module-level?

But at least for the simple case where you don't set the ns callback then you don't need to do anything.

iabdalkader replied:
I've considered this too, the only reason I went with `init()` was to allow adding more options in the future if needed, i.e. future proofing, but if that doesn't make sense init-on-demand is easy to do.

iabdalkader replied:
I removed the `init` function, now `openamp_init()` is called if it's not initialized if an `Endpoint` or a `RemoteProc` is created, or if `ns_callback()` is called. In the future, if we need to do any runtime configuration we can add it back as `openamp.config/init()` function.

dpgeorge replied:
Very good!

iabdalkader commented:
> (For my own curiosity, I'm still not entirely clear why the first way is needed though, i.e. what scenario does that support that wasn't possible the second way?)

Although probably unlikely, but a scenario which is impossible without ns_callback if you don't know the exact names, or numbers, of the endpoints the remote will announce beforehand, or if the endpoint name is random (for some reason). In this case you must create endpoints on the fly. But more commonly you may just not care about the exact name of endpoint that will be created by the remote firmware, or if it might change you don't want it to be hard-coded in the Python script (which might be a frozen module).

dpgeorge commented on extmod/libmetal/metal/config.h:
```diff
@@ -0,0 +1,46 @@
+/*
```
Can we also make this file a simple one-liner that includes `metal/micropython.h`?  Then all the configuration is in one file.

Another reason for this is so we don't have non-MIT code in this `extmod` directory.  All code currently in `extmod` is first-party code and MIT licensed.  We should try to keep it that way.

iabdalkader replied:
We can just move everything to `extmod/libmetal/metal/config.h`, see updated files.

dpgeorge replied:
Great!

dpgeorge committed:
py/stream: Factor stream implementations.

So there's only one location that does the ioctl(MP_STREAM_SEEK) call.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I pushed a commit to this branch which refactors the stream seek code, and now the coverage passes because the new code is tested by existing tests.

dpgeorge commented:
@iabdalkader can you please add some initial documentation for the new `openamp` module introduced here?

iabdalkader commented:
> @iabdalkader can you please add some initial documentation for the new `openamp` module introduced here?

I can try, but the API is not finalized yet. We need to decide what to do with `init()` function it's been discussed in one of the review comments. Also, now that `RPMsg` was renamed to `Endpoint` I think `RProc` maybe should also be renamed to `Remote` or `RemoteProc` @jimmo any thoughts on this ?

jimmo commented:
> I think `RProc` maybe should also be renamed to `Remote` or `RemoteProc` @jimmo any thoughts on this ?

@iabdalkader Yes this sounds good. Thanks

iabdalkader commented:
> I pushed a commit to this branch which refactors the stream seek code, and now the coverage passes because the new code is tested by existing tests.

@dpgeorge Thanks! But for some reason there are more failures now. Note I rebased on master and squashed the last commit.

dpgeorge commented:
> I can try, but the API is not finalized yet.

I think it's still worth starting on the docs, at least to have a skeleton there with the current API as it stands.  That will give a good summary/overview of the new modules/functions/classes/methods.

dpgeorge commented on py/stream.c:
```diff
@@ -545,16 +555,8 @@ ssize_t mp_stream_posix_read(void *stream, void *buf, size_t len) {
-    }
-    return seek_s.offset;
+    int error;
+    return mp_stream_seek(MP_OBJ_FROM_PTR(stream), offset, whence, &error);
```
It may be that you need to convert the type `mp_off_t` to `off_t` to get this working.  Something like this:
```c
mp_off_t res = mp_stream_seek(...);
if (res == (mp_off_t)-1) {
    return -1;
}
return res;
```

iabdalkader replied:
No it didn't fix it. It seems to be expecting a different errorno ?

```
OSError(2,)
```
Exp:
```
OSError(1001,)
```

iabdalkader committed:
docs/library/openamp: Document the new openamp module.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
@dpgeorge Docs added and all CI issues fixed.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+Functions
+---------
+
+.. function:: ns_callback(ns_callback)
```
I think this function should be named differently, at the very least to include the words "new service" in it, because "ns" doesn't mean much to someone new to OpenAMP (or is it "name service"?).

Maybe `new_service_callback()`?

Or `callbacks(new_service_cb)` which leaves open the possibility for additional global-scope callbacks in the future (probably there won't be any, so maybe that's over engineering).

Alternatively, we could follow the `sys` module with things like `sys.displayhook` an make this an attribute:
```py
openamp.new_service_hook = my_function
```

I think that last option (an attribute) is quite a good option.

iabdalkader replied:
I'm not sure, but I leaning towards the familiar way of setting callbacks, which is `foo.bar_callback(cb)`.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+Classes
+-------
+
+.. class:: Endpoint(name, callback, src_addr=ENDPOINT_ADDR_ANY, dst_addr=ENDPOINT_ADDR_ANY)
```
I suggest calling the arguments just `src` and `dest`.  The fact they are part of the endpoint construction indicates they are addresses.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+
+   Returns True if the endpoint is ready to send (i.e., has both a source and destination addresses)
+
+.. method:: Endpoint.send(src_addr=-1, dst_addr=-1, timeout=0)
```
I suggest `src` and `dest`.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+
+   Stop the remote processor.
+
+.. method:: RemoteProc.shutdown()
```
What is the difference between `stop` and `shutdown`?  Can you call `start` after calling `stop`?

iabdalkader replied:
shutdown stops and releases all of the remote proc's resources. It should also disable power and clocks to the remote core, if possible. Note this is also used as the finaliser.

>>  Can you call start after calling stop?

This is platform dependent. For ST it's Not possible to stop or restart the m4 core, a complete system reset is required.

dpgeorge replied:
It would be good to add these points to the docs, so users know which of `stop`/`shutdown` they should use and when, and what the restrictions are.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+          it will be assigned by the library when the endpoint is bound.
+
+Methods
+-------
```
This needs to be a sub-heading of the class heading, or just leave it out altogether.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+          endpoint locally, and it will be automatically bound when the remote announces
+          the service.
+
+Classes
```
I suggest calling this section "Endpoint class" (see eg socket docs).

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,104 @@
+        - ``timeout`` Specifies the time in milliseconds to wait for a free buffer. By default
+          `send()` function is blocking.
+
+.. class:: RemoteProc(entry)
```
This needs to go in its own section, eg "RemoteProc class".

dpgeorge commented on extmod/modopenamp_rproc.c:
```diff
@@ -0,0 +1,169 @@
+ *
+ * OpenAMP's remoteproc class.
+ */
+#if MICROPY_PY_OPENAMP_RPROC
```
This should be `MICROPY_PY_OPENAMP_REMOTE_PROC` to match the class name `openamp.RemoteProc`.

iabdalkader replied:
Should we just revert the class name to `RProc` ? Otherwise the files `modopenamp_rproc`, and `mprprocport` may need to be rename too. I will add the prefix `openamp_` to the type though.

iabdalkader replied:
Everything is now renamed to `openamp_rproc_x`. Does the class name have to match ? I like `RemoteProc` better, but would still like to keep file, functions, macros names etc.. short (i.e `rproc`)

jimmo replied:
> I like RemoteProc better

I agree. We should rename the files too. e.g. `openamp_remoteproc_x`.

iabdalkader replied:
Okay, renamed everything including files @jimmo @dpgeorge

dpgeorge commented on extmod/modopenamp_rproc.c:
```diff
@@ -0,0 +1,169 @@
+    struct remoteproc rproc;
+} rproc_obj_t;
+
+const mp_obj_type_t rproc_type;
```
This should be `openamp_remote_proc_type`.

dpgeorge commented on extmod/modopenamp_rproc.c:
```diff
@@ -0,0 +1,169 @@
+typedef struct rproc_obj_t {
+    mp_obj_base_t base;
+    struct remoteproc rproc;
+} rproc_obj_t;
```
`openamp_remote_proc_obj_t`

iabdalkader commented:
@dpgeorge Since you went through API docs, is it safe to assume that the API will be stable after those changes ? Because then I can merge (and send) other PRs that use this API, while waiting for this PR to be merged.

iabdalkader committed:
stm32/boards/ARDUINO_NICLA_VISION: Enable OpenAMP.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/boards/ARDUINO_PORTENTA_H7: Enable OpenAMP.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader commented:
Rebased to fix conflicts and enabled OpenAMP for all H747 boards.

Is there anything else that needs to be done here ?

dpgeorge commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,399 @@
+};
+static MP_DEFINE_CONST_DICT(virtio_dev_locals_dict, virtio_dev_locals_dict_table);
+
+MP_DEFINE_CONST_OBJ_TYPE(
```
Can this be `static`?  It looks like it's only used in this file.

Otherwise, if it's public, please rename to `openamp_virtio_dev_type`.

dpgeorge commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,399 @@
+};
+static MP_DEFINE_CONST_DICT(endpoint_locals_dict, endpoint_locals_dict_table);
+
+MP_DEFINE_CONST_OBJ_TYPE(
```
Can this be `static`?  Otherwise rename to begin with `openamp_`.

dpgeorge commented on ports/stm32/boards/ARDUINO_GIGA/stm32h747.ld:
```diff
@@ -14,10 +14,10 @@ MEMORY
+    FLASH_FS (r)    : ORIGIN = 0x08020000, LENGTH = 128K    /* 0x08000000 -> 0x08020000 Arduino Bootloader */
+    FLASH_TEXT (rx) : ORIGIN = 0x08040000, LENGTH = 1280K   /* 0x08020000 -> 0x08040000 Flash filesystem storage */
+                                                            /* 0x08040000 -> 0x08180000 CM7 firmware */
+                                                            /* 0x08180000 -> 0x08200000 CM4 firmware */
```
I'm not sure about this flash layout, it seems that there is some wasted space for `FLASH_FS` which is unused on this board.  Or is it required by the bootloader than the application starts at `0x08040000`?

Either way, it would be good to make this clearer with named sections, eg:
```
FLASH_BL (rx)    : ORIGIN = 0x08000000, LENGTH = 128K /* Arduino bootloader */
FLASH_UNUSED (r) : ORIGIN = 0x08020000, LENGTH = 128K /* unused? */
FLASH_APP (rx)   : ORIGIN = 0x08040000, LENGTH = 1280K /* CM7 firmware */
FLASH_CM4 (rx)   : ORIGIN = 0x08180000, LENGTH = 512K /* CM4 firmware */
```

Then use `common_bl.ld` instead of `common_blifs.ld`.

And can get rid of `_micropy_hw_internal_flash_storage_ram_cache_start` and `_micropy_hw_internal_flash_storage_start`.

Is the address 0x08180000 completely fixed, eg other non-MicroPython Arduino firmware also uses this segmentation?  I'm just wondering if 512K is a good value for the size of the CM4 firmware.  It looks reasonable to me, but once this is decided then it'll be hard to change without making it hard for users.

(Same comments go for the other 2 Arduino STM boards.)

iabdalkader replied:
The bootloader on all of those boards jumps to `0x08040000`, so we can't use that sector for code space, but we're free to use it for anything else. Currently it's only good for flash FS, when the external storage is disabled, but in the future it can be repurposed for romfs for example.
I don't think there's any good reason to remove the internal FS config right now, but I don't mind it, but I would rather revisit this in a follow-up PR, because it's not really related to this one.

> Is the address 0x08180000 completely fixed, eg other non-MicroPython Arduino firmware also uses this segmentation?

Yes the Arduino firmware uses this address when the flash is configured for a 75-25 split. There's a 50-50 split that won't be supported. Note that this address is not actually used in the firmware or hard-coded, it's more of a guideline to reserve some minimum flash space for CM4.

dpgeorge replied:
> I don't think there's any good reason to remove the internal FS config right now, but I don't mind it, but I would rather revisit this in a follow-up PR, because it's not really related to this one.

OK, that's fine.

But in that case can you please update these ld files as follows (because at the moment the comments don't match the line they are describing):
```
FLASH_BL (rx)    : ORIGIN = 0x08000000, LENGTH = 128K /* Arduino bootloader */
FLASH_FS (r)     : ORIGIN = 0x08020000, LENGTH = 128K /* filesystem */
FLASH_TEXT (rx)  : ORIGIN = 0x08040000, LENGTH = 1280K /* CM7 firmware */
FLASH_CM4 (rx)   : ORIGIN = 0x08180000, LENGTH = 512K /* CM4 firmware */
```

dpgeorge commented:
> Since you went through API docs, is it safe to assume that the API will be stable after those changes ?

Yes, the API looks stable now.

dpgeorge commented on docs/library/openamp.rst:
```diff
@@ -0,0 +1,105 @@
+        - ``dest`` The destination endpoint address of the message. If none is provided,
+          the destination address the endpoint is bound to is used.
+        - ``timeout`` Specifies the time in milliseconds to wait for a free buffer. By default
+          `send()` function is blocking.
```
Does the default of `timeout=0` mean no blocking, or infinite blocking?  If the default is blocking forever then maybe the default value here should be `timeout=-1`.

iabdalkader replied:
infinite blocking, and I've changed it to `-1` and signed int.

dpgeorge commented on extmod/modopenamp.c:
```diff
@@ -0,0 +1,399 @@
+        if (bytes > 0) {
+            break;
+        }
+        if (timeout > 0 && (mp_hal_ticks_ms() - start > timeout)) {
```
I suggest `timeout` is signed `mp_int_t` and the first condition here is `timeout >= 0`, so that a negative number means wait forever.

iabdalkader replied:
Timeout of 0 or -1 will be both be blocking, do you mean break on `timeout == 0` like so:
```C
    for (mp_uint_t start = mp_hal_ticks_ms(); ;) {
        bytes = rpmsg_send_offchannel_raw(&self->ep, src, dest, rbuf.buf, rbuf.len, false);
        if (bytes > 0) {
            break;
        }
        if (timeout == 0) {
            break;
        }
        if (timeout > 0 && (mp_hal_ticks_ms() - start > timeout)) {
            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("timeout waiting for a free buffer"));
        }
        MICROPY_EVENT_POLL_HOOK
    }
```
I think I should also return the number of bytes sent...

iabdalkader replied:
One minor change to fix a corner case, events need to be polled at least once to make sure the rproc notify node is scheduled, if send is called in a tight loop with no sleep.

iabdalkader commented:
Pushed a minor fix to detect if the CM4 is configured for auto-boot, via option bytes, and handle that case gracefully. Note in this case the CM4 can't be started, and options bytes must be disabled first (using ST tools or Arduino IDE or something else).

dpgeorge commented on ports/stm32/mpu.h:
```diff
@@ -94,6 +95,18 @@
         | MPU_REGION_ENABLE << MPU_RASR_ENABLE_Pos \
     )
 
+#define MPU_CONFIG_SHARED_NCACHE(size) ( \
```
I suggest calling this `MPU_CONFIG_SHARED_UNCACHED`, to match other names here.

iabdalkader replied:
I actually added the first one, but other code was merged first. I used `NCACHE` so we can then add `NSHARED_NCACHED_NBUFFERED` if needed, just a single letter to negate configs, vs `NOTSHARED_UNCACHED`. I'll change it anyway.

dpgeorge commented on extmod/libmetal/libmetal.mk:
```diff
@@ -0,0 +1,48 @@
+# These files are generated by the rule above (along with config.h), so we need to make the .c
+# files depend on that step -- can't use the .o files as the target because the .c files don't
+# exist yet.
+$(SRC_LIBMETAL_C): $(BUILD)/openamp/metal/config.h
```
This doesn't work as expected, because the qstr preprocessing can run before these source files are generated, and then the build fails (because qstr preprocessing includes these config files).

In particular this PR does not currently build from a fresh start.

The following change fixes it for me:
```
QSTR_GLOBAL_REQUIREMENTS += $(BUILD)/openamp/metal/config.h
```
That will also require updating the comment just above.

iabdalkader replied:
This builds fine locally, and in the CI. I'm not sure why it's not working for you, but I'll make that change anyway.

dpgeorge replied:
Is there a build in the CI that builds openamp/libmetal?  I couldn't find one.

iabdalkader replied:
You're right I thought GIGA was in the CI, should I added ? It would be a good test since it covers many features. Note it still builds here locally.

dpgeorge replied:
Yes, it would be good to add one of the stm32 Arduino boards to the CI.  You can add that in this PR as a separate commit.

(Maybe you want to add the CI first, before changing this qstr line, to see if the qstr problem is there on the CI.)

iabdalkader replied:
Yes, and if it doesn't build I'll add the line. Should it be in `ci_stm32_pyb_build` or `ci_stm32_nucleo_build` Or a new bash function like `ci_stm32_misc_build` ?

dpgeorge replied:
I think we need a new one, because the stm32 build is already taking a while.

dpgeorge replied:
Hmm, OK, let me try to work out why it doesn't work for me.

dpgeorge replied:
The issue arises when you build without `-j`, ie just a plain `make`.  I tested on the CI and indeed it also fails when built with just `make`, see https://github.com/micropython/micropython/actions/runs/8292188756/job/22693081383?pr=12961

iabdalkader replied:
Yes it fails locally too. Adding that line above, just at the end of `libmetal.mk` seems to fix.

iabdalkader committed:
tools/ci.sh: Add Arduino GIGA to stm32 CI build.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dpgeorge commented:
Note: I'm working on this PR and will merge it shortly.

dpgeorge committed:
LICENSE: Add libmetal and open-amp to 3rd-party license list.

Signed-off-by: Damien George <damien@micropython.org>

iabdalkader commented on extmod/modopenamp_remoteproc.c:
```diff
@@ -0,0 +1,175 @@
+
+// Port-defined remote-proc operations.
+const struct remoteproc_ops openamp_remoteproc_ops = {
+    .init = mp_openamp_remoteproc_init,
```
@dpgeorge do we have to keep the `mp_` prefix here ? and in the port's implementation.

dpgeorge replied:
I think it's fine having `mp_` prefix, because it's a global symbol.  It's consistent with, eg, all the functions in `ports/stm32/mpnimbleport.c`.

iabdalkader replied:
Okay no problem.

iabdalkader commented:
I reviewed the commits again, everything seems to be in order.

dpgeorge commented:
Merged!!

Thanks @iabdalkader for implementing this, I think it will be a great feature for multicore SoCs, of which there are more and more these days.
